menhir-recover/recovery_new.ml

```diff
@@ -0,0 +1,206 @@
+(*
+ * SPDX-FileCopyrightText: 2021 Serokell <https://serokell.io/>
+ * SPDX-License-Identifier: MPL-2.0
+ *)
+
+open MenhirSdk.Cmly_api
+open Synthesis
+
+module type RECOVERY = sig
+  module G : GRAMMAR
+
+  type item = G.lr1 * G.production * int
+
+  type recovery = {
+    cases: (G.lr1 option * item list) list;
+  }
+
+  val recover : G.lr1 -> recovery
+end
+
+module Recover (G : GRAMMAR) (S : SYNTHESIZER with module G := G)
+  : RECOVERY with module G := G =
+struct
+  open G
+  open Utils
+
+  type item = lr1 * production * int
+
+  type recovery = {
+    cases: (G.lr1 option * item list) list;
+  }
+
+  type trace = { cost : Cost.t; items : item list }
+
+  module Trace = struct
+    type t = trace
+    let min tr1 tr2 =
+      Cost.arg_min (fun t -> t.cost) tr1 tr2
+
+    let cat tr1 tr2 =
+      { cost = Cost.add tr1.cost tr2.cost; items = tr1.items @ tr2.items }
+  end
+
+  (* Returns the biggest `pos` among items *)
+  let depth = Lr1.tabulate (fun st ->
+    List.fold_left (fun pos (_, pos') -> max pos pos') 0 @@ Lr0.items (Lr1.lr0 st))
+
+  (*`State` is priority queue of nonterminals sorted by the cost of the trace.
+  The set stores tuples of trace cost and nonterminal. The array stores
+  traces where index is nonterminal id *)
```

Comment:
```suggestion
  (* Priority queue of nonterminals sorted by the cost of the trace. *)
```
Nitpick, but the 2nd and 3rd sentence about implementation so maybe it's better to move it in the definition of the `type t`

Reply:
Fixed